server/auth/gatekeeper.go

```diff
@@ -187,23 +210,71 @@ func (s gatekeeper) getClients(ctx context.Context, req interface{}) (*servertyp
 		if err != nil {
 			return nil, nil, status.Error(codes.Unauthenticated, err.Error())
 		}
+		if s.ssoIf.IsImpersonateEnabled() && s.ssoIf.IsRBACEnabled() {
+			return nil, nil, status.Error(codes.InvalidArgument, "only one of `sso.impersonate.enabled` or `sso.rbac.enabled` can be true")
+		}
+		if s.ssoIf.IsImpersonateEnabled() {
+			clients, err := s.impersonateAuthorization()
+			if err != nil {
+				log.WithError(err).Error("failed to setup Impersonation authorization")
+				return nil, nil, status.Error(codes.PermissionDenied, "not allowed")
+			}
+			return clients, claims, nil
+		}
 		if s.ssoIf.IsRBACEnabled() {
 			clients, err := s.rbacAuthorization(claims, req)
 			if err != nil {
 				log.WithError(err).Error("failed to perform RBAC authorization")
 				return nil, nil, status.Error(codes.PermissionDenied, "not allowed")
 			}
 			return clients, claims, nil
-		} else {
-			// important! write an audit entry (i.e. log entry) so we know which user performed an operation
-			log.WithFields(addClaimsLogFields(claims, nil)).Info("using the default service account for user")
-			return s.clients, claims, nil
 		}
+		// important! write an audit entry (i.e. log entry) so we know which user performed an operation
+		log.WithFields(addClaimsLogFields(claims, nil)).Info("using the default service account for user")
+		return s.clients, claims, nil
 	default:
 		panic("this should never happen")
 	}
 }
 
+func (s *gatekeeper) impersonateAuthorization() (*servertypes.Clients, error) {
+	restConfig := s.restConfig
+	restConfig.Wrap(
+		func(rt http.RoundTripper) http.RoundTripper {
+			return NewImpersonateRoundTripper(rt)
+		},
+	)
+
+	dynamicClient, err := dynamic.NewForConfig(restConfig)
```

Comment:
Possibly, but I would also have to refactor `ClientForAuthorization`, which I don't want to do (trying to reduce the amount of things this PR changes, as its already hard to review).

Reply:
please refactor

Reply:
@alexec I really think this is unnecessary and would make this PR much larger than it needs to be (with a high likelihood of breaking something else, as many things depend on the current signature of the `gatekeeper.clientForAuthorization` method)

If you really want, I will do it, but I think it will make reviewing this PR harder.